### ISSUE TITLE
Routed IPSec IPv6 interfaces always /128 netmask

### ISSUE THREAD

**Issue by smalchow** [opnsense/core#4700]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues and I am convinced that mine is new.

**Describe the bug**

When creating routed IPSec Tunnels, the interfaces are always created with a /128 prefix length, which already makes the interface on the other side (point-to-point with an actual netmask of /126) unreachable.

In this case it is an AWS VPN (to an AWQS Transitgateway VPN) with IPv6. 

**To Reproduce**

1. Go to 'VPN -> IPSEc -> Tunnel Settings'
2. Create a Phase 1 entry
3. Create a phase 2 entry
- mode = Route-based
- Local address, fd00::23:2
- remote address: fd00::23:1
4. on the CLI check prefix length wich is /128 and there for the remote IP is not in the subnet of the local interface

**Expected behavior**

Have the ability to configure a netmask, e.g. in pfsense you can do the in the exact same place (see screenshot)

**Describe alternatives you considered**

In regards to the BGPd we can use multi-hop which allows the bgpd to use the recieved route, but still the routing will not work

**Screenshots**

![Screenshot 2021-02-12 at 22 39 16](https://user-images.githubusercontent.com/8308427/107826006-673b9200-6d84-11eb-86fc-f84df10a6dc3.png)

**Relevant log files**

none

**Additional context**

none

**Environment**

Software version used and hardware type if relevant, e.g.:

OPNsense 21.1 (amd64, OpenSSL).
Intel® Core i5
virtio

**Comment by AdSchellevis** [comment]
wrong project? screenshot isn't our system

**Comment by AdSchellevis** [comment]
@smalchow can you check what the actual tunnel network size is when using pfSense? Ideally one would calculate the subnet between both addresses, but looking at their code I get the impression that they fix /30 (ipv4) and /64 (ipv6) on their end, probably for the same reason.

https://github.com/pfsense/pfsense/blob/65371889ad552e1d01016df68d22b5f3d7048d2c/src/etc/inc/ipsec.inc#L375-L378

My assumption is that FRR checks the interfaces netmask to determine if it's an accessible path, although it's a bit weird you didn't have issues on IPv4 on our end since we use /32 there as well.

For some reason I can't seem to set the /64 on the tunnel on my end in the same way it functions for IPv4. Just choosing a useable netmask is probably the easiest option, although I'm not 100% this is how this is supposed to work.

For reference, when trying to add a tunnel on ipv6 ifconfig complains:
```
# /sbin/ifconfig ipsec4 inet6 fd00::23:2/64 fd00::23:1
ifconfig: ioctl (SIOCAIFADDR): Invalid argument
```

Roughly the same on ipv4 works like a charm
```
# /sbin/ifconfig ipsec4 inet 192.168.234.1/30 192.168.234.2
```

**Comment by smalchow** [comment]
I will setup a lab for that tomorrow and test it, no problem

**Comment by smalchow** [comment]
Hi Ad

So, first of all, the net mask is actually a /126

[cid:199D6C3A-039C-48AC-A792-225054CEC4A1]


if you look at the pesende code in row 386 you find the relevant section, I would choose “network” for this routed tunnel.


Cheers
Soeren


On 14. Feb 2021, at 17:28, Ad Schellevis <notifications@github.com<mailto:notifications@github.com>> wrote:



@smalchow<https://github.com/smalchow> can you check what the actual tunnel network size is when using pfSense? Ideally one would calculate the subnet between both addresses, but looking at their code I get the impression that they fix /30 (ipv4) and /64 (ipv6) on their end, probably for the same reason.

https://github.com/pfsense/pfsense/blob/65371889ad552e1d01016df68d22b5f3d7048d2c/src/etc/inc/ipsec.inc#L375-L378

My assumption is that FRR checks the interfaces netmask to determine if it's an accessible path, although it's a bit weird you didn't have issues on IPv4 on our end since we use /32 there as well.

For some reason I can't seem to set the /64 on the tunnel on my end in the same way it functions for IPv4. Just choosing a useable netmask is probably the easiest option, although I'm not 100% this is how this is supposed to work.

For reference, when trying to add a tunnel on ipv6 ifconfig complains:

# /sbin/ifconfig ipsec4 inet6 fd00::23:2/64 fd00::23:1
ifconfig: ioctl (SIOCAIFADDR): Invalid argument


Roughly the same on ipv4 works like a charm

# /sbin/ifconfig ipsec4 inet 192.168.234.1/30 192.168.234.2

**Comment by AdSchellevis** [comment]
@smalchow can you post the (redacted) output of ifconfig for the ipsec device?

**Comment by AdSchellevis** [comment]
ok, that is interesting, can you try to configure an ipv4 tunnel as well? the remote address (fd00::24:1) doesn't seem to be listed in this setup, setting an address without a destination will likely do the same on our end as well, but in all examples I found online (ipv4) there always seem to be a destination involved. 

They seem to just ignore the specified destination (I can't find documentation about the reasons for this to be honest, it sounds a bit silly asking for a destination and then ignore it)
https://github.com/pfsense/pfsense/blob/65371889ad552e1d01016df68d22b5f3d7048d2c/src/etc/inc/interfaces.inc#L1701-L1703

**Comment by smalchow** [comment]
Hi,

ok here the full set, i create IPv6 and IPv4 Tunnels on opnsense and on pfsense, and the result was, that I did not fully test the tunnels for IPv4 before, they are not correct either

**opnsense**

![Screenshot 2021-02-16 at 20 48 28](https://user-images.githubusercontent.com/8308427/108113858-53e83980-7098-11eb-8cb7-c7b54d12bfa1.png)


**pfsense**

![Screenshot 2021-02-16 at 20 49 06](https://user-images.githubusercontent.com/8308427/108113931-6a8e9080-7098-11eb-9370-443f2c46169a.png)

On AWS side they were both created exactly the same way, this time i did not use short IPv6 addresses, but automatically generated but the prefix length is the interesting part anyways.

Dont mind the non-redacted IPs in the screeenshots, it is a lab only anyways which will be deleted again. 

The major difference in the configuration of those two is (this obviously applies to all tunnels)

**opnsense**

![Screenshot 2021-02-16 at 20 52 17](https://user-images.githubusercontent.com/8308427/108114718-762e8700-7099-11eb-8e81-32dd4d46981a.png)


**pfsense**

![Screenshot 2021-02-16 at 20 53 55](https://user-images.githubusercontent.com/8308427/108114780-87779380-7099-11eb-9d3f-659a35c18440.png)

Interestingly in pfsense the following happens

- if i choose "network" with a /30 (/126) prefix length, then it works correctly and uses the correct one on the interfaces.
- when i go back into the configuration it will show "address" though
- if i choose "network" and /32 the netmask will be set accordingly also on the interfaces
- if i go into the configuration again don't do anything but only save, then it is set to address and the netmask will be /30 again that seems to be the default when using address (in IPv6 the default is /64, NOT /126)

So in pfsense the code has defaults, only if you choose "network" and basically override it, then you get the prefix length you have chosen. personally i think it is questionable to say "address" and do something else, but that might only be my opinion.

You can also not derive the prefix length form the addresses you put in, even if they might fit into a /30 or /126, you still might want to set them differently. i think an explicit configuration "i want this netmask on the interface" seems to be the best.

**Comment by smalchow** [comment]
By the way, here is the code in opnsense

https://github.com/opnsense/core/blob/74ab171e1d4451bc285f0ec2a197b356cffd163c/src/etc/inc/plugins.inc.d/ipsec.inc#L567

the /32 and /128 is hardcoded, probably because the GUI elements are missing and therefore you can not get the prefix length

**Comment by AdSchellevis** [comment]
For the tunnel it usually doesn’t mind, since the other end is reachable anyway. I guess FRR assumes it can’t reach the other end.

the /32 is just as being used in the examples, it’s just odd frr doesn’t like the ipv6 tunnel since it’s technically the same. 

Thanks for the output, I’ll try to investigate a bit more later. I assume that frr also works if only both tunnel addresses are within the same range, which would be a /127 in your case.

**Comment by AdSchellevis** [comment]
@smalchow can you try https://github.com/opnsense/core/commit/8b62109a612a256c522344e0c978acdb6fefab2c?

```
opnsense-update  8b62109
```

on the latest OPNsense version should probably work, then reboot to make sure the previous devices are gone.

**Comment by smalchow** [comment]
i tried and it does create interfaces with a /127 netmask, it does not work though.

First i do not get routes at all, second i can not even ping the opposite interface anymore

Previous

![Screenshot 2021-02-17 at 17 54 19](https://user-images.githubusercontent.com/8308427/108239922-a5e99780-714a-11eb-800f-b49be62199dd.png)

Patched

![Screenshot 2021-02-17 at 17 53 11](https://user-images.githubusercontent.com/8308427/108239971-b0a42c80-714a-11eb-8e50-9abb522fc02e.png)

Interface previous

![Screenshot 2021-02-17 at 17 52 50](https://user-images.githubusercontent.com/8308427/108240018-bb5ec180-714a-11eb-9cf4-b333b1c30a63.png)

Interface now

![Screenshot 2021-02-17 at 17 52 41](https://user-images.githubusercontent.com/8308427/108240043-c154a280-714a-11eb-883f-8db29db1f425.png)

These routes I have with the /128 but with the /127 they are just gone

![Screenshot 2021-02-17 at 18 06 22](https://user-images.githubusercontent.com/8308427/108240160-db8e8080-714a-11eb-9507-f36b5462b405.png)

**Comment by AdSchellevis** [comment]
@smalchow ok, can you try the following on the command line?

```
ifconfig ipsec1 inet6 fd00::24:2 delete
ifconfig ipsec1 inet6 fd00::24:2/126
```

just a quick check if we should increase the calculated net by 1 bit.

**Comment by smalchow** [comment]
Hi,

yes, this works, no problem and i have a good route

![Screenshot 2021-02-17 at 20 27 12](https://user-images.githubusercontent.com/8308427/108257029-88262d80-715e-11eb-8d37-b24b7bc53c3f.png)

thats the way it should be.

I check linux systems we have, also running strongswan and frr, those have extra scripts (custom) to bring up the interfaces with the correct prefix, and that seems to be the only different

**Comment by AdSchellevis** [comment]
@smalchow my mistake, the 127 would only work with ::0 and ::1, https://github.com/opnsense/core/commit/ee002053da7c09b973d5e3e7c0b9a2dc0b322c8e makes sure we start at /31 and /127. you can append this patch using `opnsense-patch ee00205` when the other one is still installed. can you retest and confirm so we can close this case?

**Comment by smalchow** [comment]
i returned to a clean version and now added both patches again, reboot, and everything looks fine, the prefixes are ok and the routes are ok as well.

We will do some more testing next week, but to me it looks as if this did the trick.

@AdSchellevis thank you so much

**Comment by AdSchellevis** [comment]
@smalchow thanks for letting us know, can you close this issue when your finished testing. after a couple of release rounds we'll try to add this to stable.

**Comment by smalchow** [comment]
yes, i will

**Comment by fichtner** [comment]
@smalchow friendly ping :)

**Comment by fichtner** [comment]
Released in 21.1.3... no problems, nothing else to do: close :)
